README
5DPrint Firmware
===================

This is the firmware that powers the Makible printers (currently, the Makibox
A6).  It is based on the Sprinter firmware for RepRap-derivative controllers,
which is itself based on the Tonokips firmware.

Licensed under the GNU GPLv3.
See the LICENSE file or visit http://www.gnu.org/licenses/ for the latest 
version of the license and to find older versions of the GNU GPL licenses.

Copyright (c) 2012-2014 by Makible Limited.
 
This file is part of the 5DPrint Firmware.
 
5DPrint Firmware is free software: you can redistribute it and/or modify it 
under the terms of the GNU General Public License as published by the Free 
Software Foundation, either version 3 of the License, or (at your option) any 
later version.
 
5DPrint Firmware is distributed in the hope that it will be useful, but WITHOUT ANY 
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with 
the 5DPrint Firmware.  If not, see <http://www.gnu.org/licenses/>.

src/history.txt

Makibox A6 Firmware - History / Change Log
==========================================

Please add most recent change logs / history to the top of the list.
Save as plain text file so that no special tools are required to view this file.

--------------------------
DATE: 	19 May 2014
AUTHOR:	CJM LAI
	Makible Limited
	www.makibox.com
--------------------------
Version 2.20.42

The derivative control for the hot end is updated to give a better performance.
A multiplier of 5 is added and the low pass filter parameters are adjusted.

--------------------------
DATE: 	16 May 2014
AUTHOR:	CJM LAI
	Makible Limited
	www.makibox.com
--------------------------
Version 2.20.41

- Update PID Temperature control on hot end

  The derivative term was sampling too often and therefore, cannot detect significant
  change in temperature. The derivative term was always smaller than 1, meaning
  whatever value your PID_Kd is, it will not affect the PID control. Therefore, 
  the derivative control is changed to sample every 2 seconds and a low pass filter
  is added.
  
  Also, the functional range for integral action is reduced to 10 to reduce integral 
  windup.

- Update PID Autotune (M303) for continuous sampling as the original low pass filter 
  introduces a lag in the autotune sequence

--------------------------
DATE: 	02 May 2014
AUTHOR:	CJM LAI
	Makible Limited
	www.makibox.com
--------------------------
Version 2.20.40

- Fix PID Autotune (M303) function, main change includes multiplying 
  PIDAT_Ku by 2 to double the value of PIDAT_Ku.
- Disable Timer 1C during PID Autotune for autotune to function properly.
  The heater protection functionality is called in the autotune loop for safety reasons.
- Extra input parameter for M303, including no. of cycles and output format
- Restructure the file system
- Align M503 command output

--------------------------
DATE: 	11 March 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.34

- Limit motor current to 2A in M906 command

--------------------------
DATE: 	6 March 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.33

- Enable microstep resolution to be changed
- Add hardware information when version is printed

--------------------------
DATE: 	28 February 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.32

Add new command M908 to control GPIO pins
GPIO pins are configured as output and driven low by default.

--------------------------
DATE: 	27 February 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.31

- Update license information
- Update sdcard repository from GPLv2 to GPLv3
- Change from makibox firmware to 5D Print firmware

--------------------------
DATE: 	25 February 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.30

Main differences from version 2.20.20
- Autoprint from SD card enabled by default,
  Additional setting for enabling autoprint in EEPROM
  Additional M Command (M31) to enable/disable autoprint
- USB flashing when communicating with host
- Additional flag for current sense resistance in M906 command
- Multi-line commands with comma as the separation character

--------------------------
DATE: 	20 February 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.23


Set default sense resistance in firmware for the motor drive 
from 0.12 ohm to 0.10 ohm

Manufacturing problem:
90% => on-board sense resistance is 0.1 ohm  
10% => on-board sense resistance is 0.12 ohm

Rated current = 1700 mA

Before v2.20.23,
Rs = 0.1 ohm  => 1600 mA translates to 1920 mA *** 220 mA over the rated current ***
Rs = 0.12 ohm => 1600 mA translates to 1600 mA

After v2.20.23,
Rs = 0.1 ohm  => 1600 mA translates to 1600 mA
Rs = 0.12 ohm => 1600 mA translates to 1333 mA *** 367 mA under the rated current ***       

--------------------------
DATE: 	12 February 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.22

- In usb.c, change gnu_printf to printf for serial print out
- Enable autoprint by default
- New command M31 to turn enable / disable autoprint from SD card.
  The setting can be saved to EEPROM with the M500 command.
- Multiple commands on the same line implemented with comma as the separation character

--------------------------
DATE: 	27 January 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.21

- Update bootloader jump address according to Atmel datasheet
  Changed from 0x1E000 to 0x 1F000 (byte address)
- Bootloader size is 2048 words, word address is 0xF800 - 0xFFFF

--------------------------
DATE: 	27 January 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.20

- Furhter bug fixing for tone.c and tone.h
- New command M305 to set duty cycle of hot bed before or after 70C 
Settings for hot bed duty cycle can be changed by user and stored in EEPROM
By default, the duty cycles before and after 70C are 70% and 85% respecitvely.
These duty cycles allow a 0.7 ohm hot bed to have a maximum current about 10.8 A.
The limitation is done for safety reason, to prevent the hot bed from 
excedding the rated current limit of the power supply.

--------------------------
DATE: 	22 January 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.10

Version shipped with first batch of 5DPD8 boards
- Autoprint from SD card feature added, but disabled by default
- Start of Doxygen documentation
- Bug fixing for tone generation with buzzer
  (Tone created still has pulses due to the ISRs messing the timing)

--------------------------
DATE: 	14 January 2014
AUTHOR:	CJM LAI
	Makible Limited	
	www.makibox.com
--------------------------
Version 2.20.00


src/pins.h
/*
  5DPrint Firmware
  Based on Sprinter (master branch, 1 Sep 2012).
  Designed for Printrboard (Rev B) and 5DPrint D8 Dirver Board.
  ---
  Copyright (c) 2012-2014 by Makible Limited.
 
  This file is part of the 5DPrint Firmware.
 
  5DPrint Firmware is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
 
  The 5DPrint Firmware is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License
  along with the 5DPrint Firmware.  If not, see <http://www.gnu.org/licenses/>.
*/
/**
   \file pins.h
   \brief Pin configuration for 5DPD8 or Printrboard Rev. B
   
 */

#ifdef _5DPD8

/****************************************************************************************
 * Makibox 5DPD8 v0.12 pin assingments (ATMEGA90USB1286)
 * See http://5Dprint.com/ for more info
 ****************************************************************************************/

#define ALARM_PIN          -1

#define X_STEP_PIN         28 // PINA0
#define X_DIR_PIN          29 // PINA1
#define X_ENABLE_PIN       17 // PINC7
#define X_MS1_PIN          25 // PINB5
#define X_MS2_PIN          26 // PINB6
#define X_MIN_PIN          37 // PINE5
#define X_MAX_PIN          -1

#define Y_STEP_PIN         30 // PINA2
#define Y_DIR_PIN          31 // PINA3
#define Y_ENABLE_PIN       13 // PINC3
#define Y_MS1_PIN          9  // PINE1
#define Y_MS2_PIN          8  // PINE0
#define Y_MIN_PIN          36 // PINE4
#define Y_MAX_PIN          -1

#define Z_STEP_PIN         32 // PINA4
#define Z_DIR_PIN          33 // PINA5
#define Z_ENABLE_PIN       12 // PINC2
#define Z_MS1_PIN          7  // PIND7
#define Z_MS2_PIN          6  // PIND6
#define Z_MIN_PIN          19 // PINE7
#define Z_MAX_PIN          -1

#define E_STEP_PIN         34 // PINA6
#define E_DIR_PIN          35 // PINA7
#define E_ENABLE_PIN       11 // PINC1
#define E_MS1_PIN          5  // PIND5
#define E_MS2_PIN          4  // PIND4

#define HEATER_0_PIN       15 // PINC5, Extruder
#define HEATER_1_PIN       14 // PINC4, Bed
#define FAN_PIN            16 // PINC6, Fan

#define TEMP_0_PIN          1 // ADC0, Extruder (ADC input number)
#define TEMP_1_PIN          0 // ADC1, Bed (ADC input number)

#define LED_PIN            -1
#define USB_LED_PIN        24 // PINB4
#define PS_ON_PIN          -1
#define KILL_PIN           18 // PINE6

#define BUZZER_PIN         10 // PINC0      

#define SDCD_PIN           47 // PINE3
#define SDCS_PIN           20 // PINB0
#define SCK_PIN            21 // PINB1
#define MISO_PIN           23 // PINB3
#define MOSI_PIN           22 // PINB2

#define I2C_SCL			    0
#define I2C_SDA			    1
#define DIGIPOT_A0		   -1
#define DIGIPOT_A1		   -1
//#define DIGIPOT_RESET		    2

// GPIO pins
#define GPIO_1 40 // PINF2
#define GPIO_2 41 // PINF3
#define GPIO_3 42 // PINF4 (JTAG)
#define GPIO_4 43 // PINF5 (JTAG)
#define GPIO_5 27 // PINB7 
#define GPIO_6 44 // PINF6 (JTAG)
#define GPIO_7 45 // PINF7 (JTAG)
#define GPIO_8 2 // PIND2
#define GPIO_9 3 // PIND3

#endif // MAKIBOX_5DPD8


/****************************************************************************************
* Printrboard Rev. B pin assingments (ATMEGA90USB1286)
* Requires the Teensyduino software with Teensy2.0++ selected in arduino IDE!
* See http://reprap.org/wiki/Printrboard for more info
****************************************************************************************/

#ifdef PRINTRBOARD_REVB

#define ALARM_PIN          -1
#define X_STEP_PIN         28
#define X_DIR_PIN          29
#define X_ENABLE_PIN       19
#define X_MIN_PIN          47
#define X_MAX_PIN          -1

#define Y_STEP_PIN         30
#define Y_DIR_PIN          31
#define Y_ENABLE_PIN       18

// config.h must be included before this file
// Hopefully, should be able to remove this if-else once either new hardware is 
// release or we instruct users to use the E-STOP headers on the PrintRBoard 
// for the Y-STOP / limit switch.
#if SDSUPPORT > 0
	#define Y_MIN_PIN          37
#else
	#define Y_MIN_PIN          20
#endif

#define Y_MAX_PIN          -1

#define Z_STEP_PIN         32
#define Z_DIR_PIN          33
#define Z_ENABLE_PIN       17
#define Z_MIN_PIN          36
#define Z_MAX_PIN          -1

#define E_STEP_PIN         34
#define E_DIR_PIN          35
#define E_ENABLE_PIN       13

#define HEATER_0_PIN       15  // Extruder
#define HEATER_1_PIN       14  // Bed
#define FAN_PIN            16  // Fan

#define TEMP_0_PIN          1  // Extruder
#define TEMP_1_PIN          0  // Bed

#define LED_PIN            -1
#define PS_ON_PIN          -1
#define KILL_PIN           -1

#define SCK_PIN            21 
#define MISO_PIN           22
#define MOSI_PIN           23

#endif // PRINTRBOARD_REVB
